### PR TITLE
feat(skore)!: Add n_bins="auto" option to CalibrationDisplay

### LINKED ISSUE #2937: Add `n_bins="auto"` option to `CalibrationDisplay`
The discussion on [this scikit-learn issue](https://github.com/scikit-learn/scikit-learn/issues/33658) seems to have converged to a consensus on using the cubic rule. Let's add it to `CalibrationDisplay`.

### PR DESCRIPTION
**Pull request by direkkakkar319-ops**
#### Description
Fixes #2937 
Reference Issues : https://github.com/scikit-learn/scikit-learn/issues/33658 (for information)
Reference PR : https://github.com/scikit-learn/scikit-learn/pull/33856 (for code)

the pr adds `n_bins="auto"` option in the CalibrationDisplay report
here the default `n_bins` is also changed to `n_bins="auto"`

#### Testing code
From file path --> `F:\skore_05\skore\src\skore\_displays\inspection\calibration_curve.py`
```python
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from skore import evaluate

X, y = make_classification(
n_samples=100_000, n_features=20, n_informative=2, n_redundant=10,random_state=42)
report = evaluate(LogisticRegression(), X, y, splitter=0.2)
display = report.inspection.calibration_curve(n_bins="auto", strategy="uniform")
print(display.frame().head())
```

##### Output in `main`
(Here the `n_bins=5`)
```
(venv) direk@dkakkar:/mnt/f/skore_05$ python test_code.py
   predicted_probability  fraction_of_positives data_source  label
0               0.058169               0.058186        test      0
1               0.291218               0.279537        test      0
2               0.501687               0.507645        test      0
3               0.709106               0.705734        test      0
4               0.942051               0.940309        test      0
(venv) direk@dkakkar:/mnt/f/skore_05$ 
```

##### Output in `direkkakkar319-ops:CalibrationDisplay`
(Here the `n_bins="auto"` )
```
(venv) direk@dkakkar:/mnt/f/skore_05$ python test_code.py
   predicted_probability  fraction_of_positives data_source  label
0               0.013323               0.018338        test      0
1               0.052453               0.049888        test      0
2               0.088494               0.093787        test      0
3               0.124554               0.113264        test      0
4               0.160065               0.145329        test      0
(venv) direk@dkakkar:/mnt/f/skore_05$ 
```

### CODE REVIEW

File: skore/src/skore/_displays/inspection/calibration_curve.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ class CalibrationDisplay(DisplayMixin):
     ...     n_samples=100_000, n_features=20, n_informative=2, n_redundant=10,
     ...     random_state=42)
     >>> report = evaluate(LogisticRegression(), X, y, splitter=0.2)
-    >>> display = report.inspection.calibration_curve(n_bins=5, strategy="uniform")
+    >>> display = report.inspection.calibration_curve(n_bins="auto", strategy="uniform")
```
Comment (direkkakkar319-ops) [inline skore/src/skore/_displays/inspection/calibration_curve.py:52]: wrote n_bins="auto" is it fine
if yes then do i need to replace the shown output with the current output ?

File: skore/src/skore/_displays/inspection/calibration_curve.py [hunk 2]
```diff
@@ -144,13 +144,14 @@ def _compute_data_for_display(
...
         report_type: ReportType,
-        n_bins: int = 5,
+        n_bins: int | Literal["auto"] = 5,
         strategy: Literal["uniform", "quantile"] = "quantile",
         report_pos_label: PositiveLabel = None,
     ) -> CalibrationDisplay:
         """Compute the data for the calibration curve display."""
         y_arr = np.asarray(y)
         y_pred_arr = np.asarray(y_pred)
+        new_n_bins = int(np.ceil(len(y_arr) ** (1 / 3))) if n_bins == "auto" else n_bins
```
Comment (direkkakkar319-ops) [inline skore/src/skore/_displays/inspection/calibration_curve.py]: is the variable name `new_n_bins` fine ??

File: skore/src/skore/_displays/inspection/calibration_curve.py (hunk 1 above)
Comment (GaetandeCast) [reply skore/src/skore/_displays/inspection/calibration_curve.py:52]: Yes it is good to advertise this option in the docstring.

File: skore/src/skore/_displays/inspection/calibration_curve.py (hunk 2 above)
Comment (GaetandeCast) [reply skore/src/skore/_displays/inspection/calibration_curve.py]: I would rather you don't create a local variable and directly pass `int(np.ceil(len(y_arr) ** (1 / 3))) if n_bins == "auto" else n_bins` to `calibration_curve`.

File: skore/src/skore/_displays/inspection/calibration_curve.py [hunk 3]
```diff
@@ -144,13 +144,14 @@ def _compute_data_for_display(
         y_pred: ArrayLike,
         y: ArrayLike,
         report_type: ReportType,
-        n_bins: int = 5,
+        n_bins: int | Literal["auto"] = 5,
```
Comment (GaetandeCast) [inline skore/src/skore/_displays/inspection/calibration_curve.py]:
```suggestion
        n_bins: int | Literal["auto"] = "auto",
```
Let us put it as default.

File: skore/src/skore/_displays/inspection/calibration_curve.py (hunk 1 above)
Comment (GaetandeCast) [reply skore/src/skore/_displays/inspection/calibration_curve.py:52]: and yes you will have to replace the expected output, otherwise doctest will fail.

Comment (direkkakkar319-ops) [reply skore/src/skore/_displays/inspection/calibration_curve.py:52]: updated the docs also

File: skore/src/skore/_displays/inspection/calibration_curve.py (hunk 2 above)
Comment (direkkakkar319-ops) [reply skore/src/skore/_displays/inspection/calibration_curve.py]: now we do it directly no other variables is involved

File: skore/src/skore/_displays/inspection/calibration_curve.py (hunk 3 above)
Comment (direkkakkar319-ops) [reply skore/src/skore/_displays/inspection/calibration_curve.py]: set it to default in the other file also

File: skore/tests/unit/displays/calibration_curve/test_common.py [hunk 4]
```diff
@@ -25,6 +27,32 @@ def test_class_attributes(self, fixture_prefix, task, request):
...
+        if isinstance(report, tuple):
+            report = report[0]
+
+        observed_n_bins = []
+
+        def fake_calibration_curve(y_true, y_pred, *, n_bins, strategy):
+            expected_n_bins = int(np.ceil(len(y_true) ** (1 / 3)))
+            observed_n_bins.append(n_bins)
+            assert n_bins == expected_n_bins
+            return np.array([0.0]), np.array([0.5])
```
Comment (direkkakkar319-ops) [inline skore/tests/unit/displays/calibration_curve/test_common.py]: should we also add `assert strategy` here as it will help with just confirming the existing default behavior is still preserved.

**Comment by direkkakkar319-ops** [comment]
Hi @GaetandeCast  i think the pr is ready for your review : )

File: skore/tests/unit/displays/calibration_curve/test_common.py (hunk 4 above)
Comment (GaetandeCast) [reply skore/tests/unit/displays/calibration_curve/test_common.py]: I think this test is a bit overkill. I'd rather just get the length of the `.frame()` and assert that it is equal to the expected value when applying the cubic rule. If you set the `label=` in `.frame`, you wont have to make a separate case for binary and multiclass.

File: skore/src/skore/_displays/inspection/calibration_curve.py [hunk 5]
```diff
@@ -49,19 +49,19 @@ class CalibrationDisplay(DisplayMixin):
     ...     n_samples=100_000, n_features=20, n_informative=2, n_redundant=10,
     ...     random_state=42)
     >>> report = evaluate(LogisticRegression(), X, y, splitter=0.2)
-    >>> display = report.inspection.calibration_curve(n_bins=5, strategy="uniform")
-    >>> display.frame()
+    >>> display = report.inspection.calibration_curve(n_bins="auto", strategy="uniform")
+    >>> display.frame().head(10)
```
Comment (GaetandeCast) [inline skore/src/skore/_displays/inspection/calibration_curve.py]:
```suggestion
    >>> display.frame().head(5)
```
Nit: we don't have to print that many lines

**Review by GaetandeCast — CHANGES_REQUESTED** [review]
Rework the test to be simpler and we're good !

File: skore/tests/unit/displays/calibration_curve/test_common.py [hunk 6]
```diff
@@ -25,6 +26,17 @@ def test_class_attributes(self, fixture_prefix, task, request):
...
+    def test_auto_n_bins_uses_cube_root_rule(self, fixture_prefix, task, request):
+        report = request.getfixturevalue(f"{fixture_prefix}_{task}")
+        if isinstance(report, tuple):
+            report = report[0]
+        display = report.inspection.calibration_curve()
+        label = display.labels[0]
+        frame = display.frame(label=label)
+        n_samples = len(report.y_test)
+        expected_n_bins = int(np.ceil(n_samples ** (1 / 3)))
+        assert len(frame) <= expected_n_bins
```
Comment (GaetandeCast) [inline skore/tests/unit/displays/calibration_curve/test_common.py]:
```suggestion
        assert len(frame) == expected_n_bins
```
Any reason to not do that ?

File: skore/tests/unit/displays/calibration_curve/test_common.py [hunk 7]
```diff
@@ -25,6 +26,17 @@ def test_class_attributes(self, fixture_prefix, task, request):
         assert fig is not None
         assert len(fig.axes) >= 1
 
+    def test_auto_n_bins_uses_cube_root_rule(self, fixture_prefix, task, request):
+        report = request.getfixturevalue(f"{fixture_prefix}_{task}")
+        if isinstance(report, tuple):
+            report = report[0]
+        display = report.inspection.calibration_curve()
+        label = display.labels[0]
+        frame = display.frame(label=label)
```
Comment (GaetandeCast) [inline skore/tests/unit/displays/calibration_curve/test_common.py]:
```suggestion
        frame = display.frame(label=display.labels[0])
```

File: skore/tests/unit/displays/calibration_curve/test_common.py (hunk 6 above)
Comment (direkkakkar319-ops) [reply skore/tests/unit/displays/calibration_curve/test_common.py]: DummyClassifier(strategy="uniform") with only 20 samples, all predictions land in the same bin, so `calibration_curve` returns 1 row instead of 3(as the cubic_rule would return 3).
 


##### if we use `assert len(frame) == expected_n_bins`
image

image

if we use `assert len(frame) <= expected_n_bins`
image


but we we use logisticRegression modal it will spread the data evernly so we can use `==` but the test case was different so i used it : )

Comment (GaetandeCast) [reply skore/tests/unit/displays/calibration_curve/test_common.py]: Then let's change the model to linear and assert `==`. You can do it by adding 
```
@pytest.fixture(scope="module")
def estimator_type():
    return "linear"
```
at the top of `skore/tests/unit/displays/calibration_curve/conftest.py`.

Comment (direkkakkar319-ops) [reply skore/tests/unit/displays/calibration_curve/test_common.py]: added it at the top of the confest.py and now the test passes

**Review by GaetandeCast — APPROVED** [review]
LGTM, thanks !
